File: app/src/main/java/com/revosleap/bxplayer/Fragments/InfoFragment.java
package com.revosleap.bxplayer.Fragments;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.constraint.ConstraintLayout;
import android.support.v4.app.Fragment;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import com.revosleap.bxplayer.AppUtils.BxPlayback.BXNotificationManager;
import com.revosleap.bxplayer.AppUtils.BxPlayback.BxPlayerService;
import com.revosleap.bxplayer.AppUtils.BxPlayback.PlaybackInfoListener;
import com.revosleap.bxplayer.AppUtils.BxPlayback.PlayerAdapter;
import com.revosleap.bxplayer.AppUtils.Models.AudioModel;
import com.revosleap.bxplayer.AppUtils.Utils.AudioUtils;
import com.revosleap.bxplayer.AppUtils.Utils.EqualizerUtils;
import com.revosleap.bxplayer.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;


public class InfoFragment extends Fragment {
    @BindView(R.id.imageViewInfo)
    ImageView imageViewInfo;
    @BindView(R.id.textViewInfoTitle)
    TextView textViewInfoTitle;
    @BindView(R.id.textViewInfoArtist)
    TextView textViewInfoArtist;
    @BindView(R.id.buttonInfoPlaylist)
    Button buttonInfoPlaylist;
    @BindView(R.id.buttonInfoFave)
    Button buttonInfoFave;
    @BindView(R.id.buttonInfoVol)
    Button buttonInfoVol;
    @BindView(R.id.constraintLayout2)
    ConstraintLayout constraintLayout2;
    @BindView(R.id.seekBarInfo)
    SeekBar seekBarInfo;
    @BindView(R.id.buttonInfoShuffle)
    Button buttonInfoShuffle;
    @BindView(R.id.buttonInfoPrev)
    Button buttonInfoPrev;
    @BindView(R.id.buttonInfoPlay)
    Button buttonInfoPlay;
    @BindView(R.id.buttonInfoNext)
    Button buttonInfoNext;
    @BindView(R.id.buttonInfoAll)
    Button buttonInfoAll;
    @BindView(R.id.constraintLayout3)
    ConstraintLayout constraintLayout3;
    Unbinder unbinder;
    boolean mUserIsSeeking,mIsBound;
    @BindView(R.id.textViewProgress)
    TextView textViewProgress;
    @BindView(R.id.textViewDuration)
    TextView textViewDuration;
    private String mSelectedArtist;
    private PlayerAdapter mPlayerAdapter;
    BxPlayerService mMusicService;
    private BXNotificationManager mMusicNotificationManager;
    private PlaybackListener mPlaybackListener;
    private ServiceConnection mConnection= new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mMusicService=((BxPlayerService.BXBinder)service).getInstance();
            mPlayerAdapter = mMusicService.getMediaPlayerHolder();
            mMusicNotificationManager = mMusicService.getMusicNotificationManager();
            if (mPlaybackListener == null) {
                mPlaybackListener = new PlaybackListener();
                mPlayerAdapter.setPlaybackInfoListener(mPlaybackListener);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mMusicService = null;
        }
    };


    public InfoFragment() {
    }


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.info, container, false);
        unbinder = ButterKnife.bind(this, view);
        onViewClicked(view);
        doBindService();
        initializeSeekbar();
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }
    private void doBindService() {
        // Establish a connection with the service.  We use an explicit
        // class name because we want a specific service implementation that
        // we know will be running in our own process (and thus won't be
        // supporting component replacement by other applications).
        getActivity().bindService(new Intent(getActivity(),
                BxPlayerService.class), mConnection, Context.BIND_AUTO_CREATE);
        mIsBound = true;

        final Intent startNotStickyIntent = new Intent(getActivity(),BxPlayerService.class);
        getActivity().startService(startNotStickyIntent);
    }
    private void doUnbindService() {
        if (mIsBound) {
            // Detach our existing connection.
            getActivity().unbindService(mConnection);
            mIsBound = false;
        }
    }

    @OnClick({R.id.buttonInfoPlaylist, R.id.buttonInfoFave, R.id.buttonInfoVol, R.id.buttonInfoShuffle,
            R.id.buttonInfoPrev, R.id.buttonInfoPlay, R.id.buttonInfoNext, R.id.buttonInfoAll})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.buttonInfoPlaylist:
                break;
            case R.id.buttonInfoFave:
                break;
            case R.id.buttonInfoVol:
                break;
            case R.id.buttonInfoShuffle:
                break;
            case R.id.buttonInfoPrev:
                skipPrev();
                break;
            case R.id.buttonInfoPlay:
                resumeOrPause();
                break;
            case R.id.buttonInfoNext:
                skipNext();
                break;
            case R.id.buttonInfoAll:
                break;
        }
    }

    private void initializeSeekbar() {
        seekBarInfo.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            int userSelectedPosition = 0;

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    userSelectedPosition = progress;

                }
                textViewProgress.setText(AudioUtils.formatDuration(progress));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                mUserIsSeeking = true;
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                if (mUserIsSeeking) {
                  //  mSongPosition.setTextColor(currentPositionColor);
                }
                mUserIsSeeking = false;
                mPlayerAdapter.seekTo(userSelectedPosition);
            }
        });
    }
    private void updatePlayingStatus() {
        final int drawable = mPlayerAdapter.getState() != PlaybackInfoListener.State.PAUSED ?
                R.drawable.pause : R.drawable.play_icon;
       buttonInfoPlay.post(new Runnable() {
            @Override
            public void run() {
               buttonInfoPlay.setBackgroundResource(drawable);
            }
        });
    }
    private void updatePlayingInfo(boolean restore, boolean startPlay) {

        if (startPlay) {
            mPlayerAdapter.getMediaPlayer().start();
            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    mMusicService.startForeground(BXNotificationManager.NOTIFICATION_ID,
                            mMusicNotificationManager.createNotification());
                }
            }, 250);
        }

        final AudioModel selectedSong = mPlayerAdapter.getCurrentSong();

        mSelectedArtist = selectedSong.getArtist();
        final int duration = selectedSong.getDuration();
        seekBarInfo.setMax(duration);

        textViewDuration.setText(AudioUtils.formatDuration(duration));

        textViewInfoTitle.post(new Runnable() {
            @Override
            public void run() {
                textViewInfoTitle.setText(selectedSong.getTitle());
            }
        });

        textViewInfoArtist.setText(selectedSong.getArtist());

        if (restore) {
            seekBarInfo.setProgress(mPlayerAdapter.getPlayerPosition());
            updatePlayingStatus();
           // updateResetStatus(false);

            new Handler().postDelayed(new Runnable() {
                @Override
                public void run() {
                    //stop foreground if coming from pause state
                    if (mMusicService.isRestoredFromPause()) {
                        mMusicService.stopForeground(false);
                        mMusicService.getMusicNotificationManager().getNotificationManager()
                                .notify(BXNotificationManager.NOTIFICATION_ID,
                                        mMusicService.getMusicNotificationManager().getNotificationBuilder().build());
                        mMusicService.setRestoredFromPause(false);
                    }
                }
            }, 250);
        }
    }
    class PlaybackListener extends PlaybackInfoListener {

        @Override
        public void onPositionChanged(int position) {
            if (!mUserIsSeeking) {
                seekBarInfo.setProgress(position);
            }
        }

        @Override
        public void onStateChanged(@State int state) {

            updatePlayingStatus();
            if (mPlayerAdapter.getState() != State.RESUMED && mPlayerAdapter.getState() != State.PAUSED) {
                updatePlayingInfo(false, true);
            }
        }

        @Override
        public void onPlaybackCompleted() {
            //updateResetStatus(true);
        }
    }
    private boolean checkIsPlayer() {

        boolean isPlayer = mPlayerAdapter.isMediaPlayer();
        if (!isPlayer) {
            EqualizerUtils.notifyNoSessionId(getActivity());
        }
        return isPlayer;
    }
    public void skipPrev() {
        if (checkIsPlayer()) {
            mPlayerAdapter.instantReset();
        }
    }

    public void resumeOrPause() {
        if (checkIsPlayer()) {
            mPlayerAdapter.resumeOrPause();
        }
    }

    public void skipNext() {
        if (checkIsPlayer()) {
            mPlayerAdapter.skip(true);
        }
    }

    public void openEqualizer() {
        if (EqualizerUtils.hasEqualizer(getActivity())) {
            if (checkIsPlayer()) {
                mPlayerAdapter.openEqualizer(getActivity());
            }
        } else {
            Toast.makeText(getActivity(),"No equilizer found", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/revosleap/bxplayer/ui/fragments/InfoFragment.kt
package com.revosleap.bxplayer.ui.fragments


import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.support.v7.graphics.Palette
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.Toast
import com.revosleap.bxplayer.R
import com.revosleap.bxplayer.callbacks.BXColor
import com.revosleap.bxplayer.services.MusicPlayerService
import com.revosleap.bxplayer.ui.activities.PlayerActivity
import com.revosleap.bxplayer.utils.playback.BXNotificationManager
import com.revosleap.bxplayer.utils.playback.PlaybackInfoListener
import com.revosleap.bxplayer.utils.playback.PlayerAdapter
import com.revosleap.bxplayer.utils.utils.AudioUtils
import com.revosleap.bxplayer.utils.utils.EqualizerUtils
import kotlinx.android.synthetic.main.info.*
import org.jetbrains.anko.startService
import org.jetbrains.anko.toast
import java.io.ByteArrayInputStream
import java.io.InputStream


class InfoFragment : Fragment(), View.OnClickListener,BXColor {
    private var mUserIsSeeking: Boolean = false
    private var mSelectedArtist: String? = null
    private var mPlayerAdapter: PlayerAdapter? = null
    private var mMusicService: MusicPlayerService? = null
    private var mMusicNotificationManager: BXNotificationManager? = null
    private var mPlaybackListener: PlaybackListener? = null
    private var playerActivity:PlayerActivity?=null
    override fun onAttach(context: Context?) {
        super.onAttach(context)
        getService()
        playerActivity =activity!!as PlayerActivity
        playerActivity?.setColorCallback(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        return inflater.inflate(R.layout.info, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setClickListeners()
        initializeSeekBar()
        mPlaybackListener?.onStateChanged(PlaybackInfoListener.State.PLAYING)
    }

    override fun onResume() {
        super.onResume()
        getService()
        activity?.startService<MusicPlayerService>()
        mPlayerAdapter!!.getMediaPlayer()?.start()
        mMusicService!!.startForeground(BXNotificationManager.NOTIFICATION_ID,
                mMusicNotificationManager!!.createNotification())
    }

    override fun songColor(color: Int) {
        buttonInfoPlaylist?.setColorFilter(color)
        buttonInfoPlay?.setColorFilter(color)
        buttonInfoAll?.setColorFilter(color)
        buttonInfoFave?.setColorFilter(color)
        buttonInfoVol?.setColorFilter(color)
        seekBarInfo?.progressBackgroundTintList = ColorStateList.valueOf(color)
        buttonInfoShuffle?.setColorFilter(color)
        buttonInfoNext?.setColorFilter(color)
        buttonInfoPrev?.setColorFilter(color)
    }
    private fun getBxColor(bitmap: Bitmap){
        var color:Int
        color = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            activity?.resources!!.getColor(R.color.colorAccent, null)
        } else activity?.resources!!.getColor(R.color.colorAccent)
        Palette.from(bitmap).generate { palette ->
            val vibrant = palette?.lightVibrantSwatch
            try {
                color = vibrant?.rgb!!
               songColor(color)
                playerActivity?.setViewColors(color)
            } catch (e: Exception) {
            }
        }


    }
    private fun getService(){
        val playerActivity = activity as PlayerActivity
        mMusicService = playerActivity.getPlayerService()
        if (mMusicService != null) {
            mPlayerAdapter = mMusicService!!.mediaPlayerHolder
            mMusicNotificationManager = mMusicService!!.musicNotificationManager
        }
        if (mPlaybackListener == null) {
            mPlaybackListener = PlaybackListener()
            mPlayerAdapter?.setPlaybackInfoListener(mPlaybackListener!!)
        }
    }
    private fun initializeSeekBar() {
        seekBarInfo.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            var userSelectedPosition = 0

            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    userSelectedPosition = progress

                }
                textViewProgress?.text = AudioUtils.formatDuration(progress)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                mUserIsSeeking = true
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                if (mUserIsSeeking) {
                    //  mSongPosition.setTextColor(currentPositionColor);
                }
                mUserIsSeeking = false
                mPlayerAdapter!!.seekTo(userSelectedPosition)
            }
        })
    }

    private fun updatePlayingStatus() {
        val drawable = if (mPlayerAdapter?.getState() != PlaybackInfoListener.State.PAUSED)
            R.drawable.pause
        else
            R.drawable.play_icon
        try {
            buttonInfoPlay.setImageDrawable(activity?.getDrawable(drawable))

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun updatePlayingInfo(restore: Boolean, startPlay: Boolean) {


        if (startPlay) {
            mPlayerAdapter?.getMediaPlayer()?.start()
            Handler().postDelayed({
                mMusicService!!.startForeground(BXNotificationManager.NOTIFICATION_ID,
                        mMusicNotificationManager!!.createNotification())
            }, 250)
        }

        val selectedSong = mPlayerAdapter?.getCurrentSong()
        mSelectedArtist = selectedSong?.artist
        val duration = selectedSong?.duration
        seekBarInfo?.max = duration!!
        textViewDuration?.text = AudioUtils.formatDuration(duration)
        textViewInfoTitle?.text = selectedSong.title
        textViewInfoArtist?.text = selectedSong.artist
        val retriever = MediaMetadataRetriever()
        val inputStream: InputStream?
        retriever.setDataSource(selectedSong?.path)

        if (retriever.embeddedPicture != null) {
            inputStream = ByteArrayInputStream(retriever.embeddedPicture)
            val image =BitmapFactory.decodeStream(inputStream)
            imageViewInfo?.setImageBitmap(image)
            try {
                playerActivity?.updatePlaying(selectedSong)
                getBxColor(image)
            } catch (e: Exception) {
            }

        }else{
            val image = BitmapFactory.decodeResource(playerActivity?.resources!!,R.drawable.cover2)
            imageViewInfo?.setImageBitmap(image)
            try {
                playerActivity?.updatePlaying(selectedSong)
                getBxColor(image)
            } catch (e: Exception) {
            }
        }


        if (restore) {
            seekBarInfo?.progress = mPlayerAdapter?.getPlayerPosition()!!
            updatePlayingStatus()

        }

    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.buttonInfoPlaylist -> {
            }
            R.id.buttonInfoFave -> {
            }
            R.id.buttonInfoVol -> {
            }
            R.id.buttonInfoShuffle -> {
            }
            R.id.buttonInfoPrev -> skipPrev()
            R.id.buttonInfoPlay -> resumeOrPause()
            R.id.buttonInfoNext -> skipNext()
            R.id.buttonInfoAll -> {
            }
        }
    }

    private fun setClickListeners() {
        buttonInfoAll.setOnClickListener(this)
        buttonInfoFave.setOnClickListener(this)
        buttonInfoNext.setOnClickListener(this)
        buttonInfoPlay.setOnClickListener(this)
        buttonInfoPlaylist.setOnClickListener(this)
        buttonInfoPrev.setOnClickListener(this)
        buttonInfoShuffle.setOnClickListener(this)
        buttonInfoVol.setOnClickListener(this)
    }

    internal inner class PlaybackListener : PlaybackInfoListener() {

        override fun onPositionChanged(position: Int) {
            if (!mUserIsSeeking) {
                seekBarInfo?.progress = position
            }
        }

        override fun onStateChanged(@State state: Int) {
            updatePlayingStatus()
            if (mPlayerAdapter?.getState() != PlaybackInfoListener.State.RESUMED && mPlayerAdapter?.getState() != PlaybackInfoListener.State.PAUSED) {
                updatePlayingInfo(restore = false, startPlay = true)
            }
            if (state == PlaybackInfoListener.State.PLAYING) {
                updatePlayingInfo(restore = true, startPlay = false)

            }
        }

        override fun onPlaybackCompleted() {
            //updateResetStatus(true);
        }
    }

    private fun checkIsPlayer(): Boolean {

        val isPlayer = mPlayerAdapter?.isMediaPlayer()
        if (!isPlayer!!) {
            EqualizerUtils.notifyNoSessionId(activity!!)
        }
        return isPlayer
    }

    private fun skipPrev() {
        if (checkIsPlayer()) {
            mPlayerAdapter?.instantReset()
        }
    }

    private fun resumeOrPause() {
        if (checkIsPlayer()) {
            mPlayerAdapter?.resumeOrPause()
        }
    }

    private fun skipNext() {
        if (checkIsPlayer()) {
            mPlayerAdapter!!.skip(true)
        }
    }

    fun openEqualizer() {
        if (EqualizerUtils.hasEqualizer(activity!!)) {
            if (checkIsPlayer()) {
                mPlayerAdapter?.openEqualizer(activity!!)
            }
        } else {
            Toast.makeText(activity, "No equalizer found", Toast.LENGTH_SHORT).show()
        }
    }


}
